File: src/semantic/SymbolTable.cpp
// ═══════════════════════════════════════════════════════════════════════════
// OLang Symbol Table Implementation
// Milestone 1.2: Type System Foundation
// ═══════════════════════════════════════════════════════════════════════════

#include "olang/semantic/SymbolTable.hpp"

namespace olang::semantic {

bool SymbolTable::declare(const std::string& name, 
                          std::unique_ptr<frontend::Type> type,
                          bool isConst, bool isFunction) {
    if (existsInCurrentScope(name)) {
        return false; // Symbol already exists
    }
    
    symbols_[name] = std::make_unique<Symbol>(
        name, std::move(type), isConst, isFunction);
    return true;
}

Symbol* SymbolTable::lookup(const std::string& name) {
    auto it = symbols_.find(name);
    if (it != symbols_.end()) {
        return it->second.get();
    }
    
    // Search parent scope
    if (parent_) {
        return parent_->lookup(name);
    }
    
    return nullptr;
}

const Symbol* SymbolTable::lookup(const std::string& name) const {
    auto it = symbols_.find(name);
    if (it != symbols_.end()) {
        return it->second.get();
    }
    
    if (parent_) {
        return parent_->lookup(name);
    }
    
    return nullptr;
}

bool SymbolTable::existsInCurrentScope(const std::string& name) const {
    return symbols_.find(name) != symbols_.end();
}

void SymbolTable::enterScope() {
    // In full implementation, would push new scope
}

void SymbolTable::exitScope() {
    // In full implementation, would pop scope
}

} // namespace olang::semantic

File: src/semantic/symbol_table.rs
use crate::parser::ast::Type;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct Symbol {
    pub name: String,
    pub ty: Type,
    pub is_const: bool,
    pub is_function: bool,
}

impl Symbol {
    pub fn new(name: impl Into<String>, ty: Type, is_const: bool, is_function: bool) -> Self {
        Self { name: name.into(), ty, is_const, is_function }
    }
}

/// Scoped symbol table using a stack of hash maps.
#[derive(Debug, Default)]
pub struct SymbolTable {
    scopes: Vec<HashMap<String, Symbol>>,
}

impl SymbolTable {
    pub fn new() -> Self {
        Self { scopes: vec![HashMap::new()] }
    }

    /// Push a new lexical scope.
    pub fn enter_scope(&mut self) {
        self.scopes.push(HashMap::new());
    }

    /// Pop the innermost scope.
    pub fn exit_scope(&mut self) {
        if self.scopes.len() > 1 {
            self.scopes.pop();
        }
    }

    /// Declare a symbol in the current scope. Returns `false` if already declared.
    pub fn declare(&mut self, name: impl Into<String>, ty: Type, is_const: bool, is_function: bool) -> bool {
        let name = name.into();
        let scope = self.scopes.last_mut().expect("at least one scope");
        if scope.contains_key(&name) {
            return false;
        }
        scope.insert(name.clone(), Symbol::new(name, ty, is_const, is_function));
        true
    }

    /// Look up a symbol walking out through parent scopes.
    pub fn lookup(&self, name: &str) -> Option<&Symbol> {
        for scope in self.scopes.iter().rev() {
            if let Some(sym) = scope.get(name) {
                return Some(sym);
            }
        }
        None
    }

    /// Check current scope only.
    pub fn exists_in_current_scope(&self, name: &str) -> bool {
        self.scopes.last().map(|s| s.contains_key(name)).unwrap_or(false)
    }
}
